File: kaop/src/main/java/io/github/coyamo/kaop/AspectChain.java
package io.github.coyamo.kaop;

import android.util.Log;

import java.util.ArrayList;
import java.util.List;

class AspectChain {
    private final List<Aspect> aspects = new ArrayList<>();

    public void addAspect(Aspect aspect){
        aspects.add(aspect);
    }

    public Object advice(AbsJoinPoint point){
        AbsJoinPoint localPoint = point;
        for (Aspect plugin : aspects) {
            Object result = plugin.advice(localPoint);
            localPoint = new ProxyJoinPoint(localPoint, result);
        }
        return localPoint.proceed();
    }
}

File: kaop/src/main/java/io/github/coyamo/kaop/AspectChain.kt
package io.github.coyamo.kaop

/**
 * 链式调用执行切面
 * @author Coyamo
 * @date 2022/12/17 00:47
 * @version 1.0
 */
internal class AspectChain(private val joinPoint: JoinPoint) {
    private val aspects: MutableList<Aspect> = ArrayList()

    fun addAspect(aspect: Aspect) {
        aspects.add(aspect)
    }

    fun advice(): Any? {
        var localPoint: AbsJoinPoint = joinPoint
        for (plugin in aspects) {
            val result = plugin.advice(localPoint)
            localPoint = ProxyJoinPoint(localPoint, result)
        }
        return localPoint.proceed()
    }
}